### ISSUE TITLE
TypeError during .byType()

### ISSUE THREAD

**Issue by gmaclennan** [noffle/kappa-osm#15]
We've seen a couple of [errors in the Mapeo Mobile bug tracker](https://app.bugsnag.com/digital-democracy/mapeo-mobile/errors/5e41adf32df450001a20f011?event_id=5e41adf30055f36677ea0000&i=gh&m=ci) https://github.com/digidem/mapeo-mobile/issues/306 which seem to be caused by `elm` being `undefined` or `null` in https://github.com/digidem/kappa-osm/blob/master/index.js#L370

@noffle do you have an idea of how / why `elm` would be `undefined` here without `_getByVersion()` throwing an error? This error has only come up a couple of times, and I don't know from what device, but it would be good to figure out the root cause. We could always just to a `if (!elm) return` but it would be good to figure out what is causing it to be `undefined` in the first place.

**Comment by hackergrrl** [comment]
@gmaclennan looking at the code for [_getByVersion](https://github.com/digidem/kappa-osm/blob/master/index.js#L114), it will return `null` when the feed is not present locally. But since it was indexed, it must have existed locally.

I'm not sure under what conditions `undefined` would be returned. If the block wasn't present, `hypercore#get` would return an error.

I don't think I have login credentials to access the bug tracker link you shared. Could you get me access or paste the errors here?

**Comment by gmaclennan** [comment]
@noffle I've added you as a collaborator for the bug reporter. Not much information on their unfortunately - I need to set up better reporting for the backend part of Mapeo Mobile, the frontend has more metadata. Here is what is reported:

```
TypeError · Cannot convert undefined or null to object

/data/data/com.mapeo/files/nodejs-project/index.js:37464:24

var fetch = through.obj(function (row, _, next) {
  self._getByVersion(row.version, function (err, elm) {
    if (err) return next(err)
--> var res = Object.assign(elm, {
      version: row.version,
      deviceId: versionToDeviceId(row.version),
      id: row.id
```

**Comment by gmaclennan** [comment]
could this happen if a sync had failed before completion and the db was in sparse mode?

**Comment by hackergrrl** [comment]
@gmaclennan It shouldn't. Documents would only be indexed if they existed locally.

For now, I'll patch kappa-osm to skip missing elements. It shouldn't be crashing anyways.

Ugh this means backporting the fix back to kappa-osm@3, since @4 has the hypercore8 changes that aren't in the field yet. :'(

**Comment by hackergrrl** [comment]
Published to `4.0.2` and `3.0.2`.
